src/interface/container.cpp

```diff
@@ -147,6 +143,221 @@ Container<T> Container<T>::SparseSlice(int id) {
+}
+
+template <typename T>
+VariablePack<T> Container<T>::PackVariables_(const std::vector<std::string> &names,
+                                             const vpack_types::VarList<T> &vars,
+                                             PackIndexMap &vmap) {
+  auto kvpair = varPackMap_.find(names);
+  if (kvpair == varPackMap_.end()) {
+    auto pack = MakePack<T>(vars, &vmap);
+    PackIndxPair<T> value;
+    value.pack = pack;
+    value.map = vmap;
+    varPackMap_[names] = value;
+    // varPackMap_[names] = std::make_pair(pack,vmap);
+    return pack;
+  }
+  vmap = (kvpair->second).map;
+  return (kvpair->second).pack;
+  // vmap = std::get<1>(kvpair->second);
+  // return std::get<0>(kvpair->second);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables(const std::vector<std::string> &names,
+                                            const std::vector<int> &sparse_ids,
+                                            PackIndexMap &vmap) {
+  static std::vector<std::string> all_names;
+  vpack_types::VarList<T> vars = MakeList_(names, all_names, sparse_ids);
+  return PackVariables_(all_names, vars, vmap);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables(const std::vector<std::string> &names,
+                                            const std::vector<int> &sparse_ids) {
+  PackIndexMap vmap;
+  return PackVariables(names, sparse_ids, vmap);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables(const std::vector<std::string> &names,
+                                            PackIndexMap &vmap) {
+  return PackVariables(names, {}, vmap);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables(const std::vector<std::string> &names) {
+  PackIndexMap vmap;
+  return PackVariables(names, {}, vmap);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables(const std::vector<MetadataFlag> &flags,
+                                            PackIndexMap &vmap) {
+  static std::vector<std::string> vnams;
+  vpack_types::VarList<T> vars = MakeList_(flags, vnams);
+  return PackVariables_(vnams, vars, vmap);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables(const std::vector<MetadataFlag> &flags) {
+  PackIndexMap vmap;
+  return PackVariables(flags, vmap);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables(PackIndexMap &vmap) {
+  static std::vector<std::string> vnams;
+  vpack_types::VarList<T> vars = MakeList_(vnams);
+  return PackVariables_(vnams, vars, vmap);
+}
+template <typename T>
+VariablePack<T> Container<T>::PackVariables() {
+  PackIndexMap vmap;
+  return PackVariables(vmap);
+}
+
+template <typename T>
+vpack_types::VarList<T> Container<T>::MakeList_(const std::vector<std::string> &names,
+                                                std::vector<std::string> &names_out,
+                                                const std::vector<int> sparse_ids) {
```

Comment:
Please add some developer documentation that explains what this function does. Also, I'm not a fan of non-const references, would be better to use a pointer so that the caller understands that it can be modified.

Reply:
I prefer non-const references over pointers. I think the fact that it's not const implies it can be modified.

I will add comments explaining what this does.

Reply:
The explanation, by the way is:
```C++
// From a given container, extract all variables and all fields in sparse variables
// into a single linked list of variables. The sparse fields are then named
// variable_index.
// The names of the non-sparse variables and the sparse fields are then
// packed into the std::vector "expanded_names," which is used as the key for
// the pack cache.
```

Reply:
Comments added inline. The basic idea of this is that it takes a set of names of variables and/or metadata and expands that into a linked list of pointers to the variables *and* pulls out the sparse indexes for the sparse variables, optionally only small subset of sparse indices.

The `expanded_names` variable is the names of each variable associated with the elements of the list. Importantly, it expands out the sparse variables to be `sparse_name_idx`, where `idx` is the sparse index of each element of the sparse variable.